lib/Infer/Interpreter.cpp

```diff
@@ -368,53 +404,200 @@ namespace souper {
         return Result;
       }
     }
-    case Inst::Shl : {
-      auto Op0KB = findKnownBits(I->Ops[0], C);
-      auto Op1V = getValue(I->Ops[1], C);
-      if (Op1V.hasValue()) {
-        Op0KB.One <<= Op1V.getValue();
-        Op0KB.Zero <<= Op1V.getValue();
-        Op0KB.Zero.setLowBits(Op1V.getValue().getLimitedValue());
-        // setLowBits takes an unsiged int, so getLimitedValue is harmless
-        return Op0KB;
-      } else {
-        return Result;
-      }
-    }
-    case Inst::And : {
-      auto Op0KB = findKnownBits(I->Ops[0], C);
-      auto Op1KB = findKnownBits(I->Ops[1], C);
+//   case Phi:
+//     return "phi";
+//   case Add:
+//     return "add";
+//   case AddNSW:
+//     return "addnsw";
+//   case AddNUW:
+//     return "addnuw";
+//   case AddNW:
+//     return "addnw";
+//   case Sub:
+//     return "sub";
+//   case SubNSW:
+//     return "subnsw";
+//   case SubNUW:
+//     return "subnuw";
+//   case SubNW:
+//     return "subnw";
+//   case Mul:
+//     return "mul";
+//   case MulNSW:
+//     return "mulnsw";
+//   case MulNUW:
+//     return "mulnuw";
+//   case MulNW:
+//     return "mulnw";
+//   case UDiv:
+//     return "udiv";
+//   case SDiv:
+//     return "sdiv";
+//   case UDivExact:
+//     return "udivexact";
+//   case SDivExact:
+//     return "sdivexact";
+//   case URem:
+//     return "urem";
+//   case SRem:
+//     return "srem";
+    case souper::Inst::And : {
+      auto Op0KB = KB0;
+      auto Op1KB = KB1;
 
       Op0KB.One &= Op1KB.One;
       Op0KB.Zero |= Op1KB.Zero;
       return Op0KB;
     }
     case Inst::Or : {
-      auto Op0KB = findKnownBits(I->Ops[0], C);
-      auto Op1KB = findKnownBits(I->Ops[1], C);
+      auto Op0KB = KB0;
+      auto Op1KB = KB1;
 
       Op0KB.One |= Op1KB.One;
       Op0KB.Zero &= Op1KB.Zero;
       return Op0KB;
     }
     case Inst::Xor : {
-      auto Op0KB = findKnownBits(I->Ops[0], C);
-      auto Op1KB = findKnownBits(I->Ops[1], C);
+      auto Op0KB = KB0;
+      auto Op1KB = KB1;
       llvm::APInt KnownZeroOut =
         (Op0KB.Zero & Op1KB.Zero) | (Op0KB.One & Op1KB.One);
       Op0KB.One = (Op0KB.Zero & Op1KB.One) | (Op0KB.One & Op1KB.Zero);
       Op0KB.Zero = std::move(KnownZeroOut);
       // ^ logic copied from LLVM ValueTracking.cpp
       return Op0KB;
     }
+    case souper::Inst::Shl : {
+      auto Op0KB = KB0;
+      auto Op1V = getValue(I->Ops[1], C);
+
+      if (Op1V.hasValue()) {
+        auto Val = Op1V.getValue().getLimitedValue();
+        if (Val < 0 || Val >= I->Width) {
+          return Result;
+        }
+        Op0KB.One <<= Val;
+        Op0KB.Zero <<= Val;
+        Op0KB.Zero.setLowBits(Val);
+        // setLowBits takes an unsigned int, so getLimitedValue is harmless
+        return Op0KB;
+      } else {
+        return Result;
+      }
+    }
+//   case ShlNSW:
```

Comment:
I believe all UB-flag versions of each instruciton can just use the normal version
this'll miss a bit of precision but it's a lot better than nothing

Reply:
in other words, addnsw can maybe return a better answer than add, but the answer for add should be a sound answer for addnsw